NetKVM/Common/ParaNdis_TX.cpp

```diff
@@ -567,6 +569,44 @@ PNET_BUFFER_LIST CNBL::DetachInternalObject()
     return Res;
 }
 
+NBLHistory::NBLHistory(LPCSTR Func,
+                       LPCSTR Title,
+                       LPCSTR ParameterMeaning,
+                       PVOID Parameter,
+                       LPCSTR ValueMeaning,
+                       ULONG Value)
+{
+    m_Function = Func;
+    m_Title = Title;
+    m_ParameterMeaning = ParameterMeaning;
+    m_Parameter = Parameter;
+    m_ValueMeaning = ValueMeaning;
+    m_Value = Value;
+    m_CurrentCPU = KeGetCurrentProcessorNumber();
+    UpdateTimestamp(m_Timestamp);
+}
+
+#if NBL_MAINTAIN_HISTORY
+void CNBL::AddHistory(LPCSTR Func,
+                      LPCSTR Title,
+                      LPCSTR ParameterMeaning,
+                      PVOID Parameter,
+                      LPCSTR ValueMeaning,
+                      ULONG Value)
+{
+    auto entry = new (m_Context->MiniportHandle) NBLHistory(Func,
+                                                            Title,
+                                                            ParameterMeaning,
+                                                            Parameter,
+                                                            ValueMeaning,
+                                                            Value);
+    if (entry)
+    {
+        m_History.PushBack(entry);
```

Comment:
Are you sure the destructor will be called?

Reply:
Of course. I worked with history under verifier. CNBL destructor frees the history, see TX.cpp.